File: LeetCode/Explore/LinkList/AddTwoNumbers.cpp
#include "listnode.h"

class Solution {
  public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        ListNode *head = new ListNode(0), *p = head;
        int carry = 0;
        while (l1 != nullptr || l2 != nullptr) {
            int a = l1 != nullptr ? l1->val : 0;
            int b = l2 != nullptr ? l2->val : 0;
            int sum = a + b + carry;
            carry = sum / 10;
            p->next = new ListNode(sum % 10);
            p = p->next;
            if (l1 != nullptr)
                l1 = l1->next;
            if (l2 != nullptr)
                l2 = l2->next;
        }
        if (carry > 0)
            p->next = new ListNode(carry);
        p = head;
        head = head->next;
        delete p;
        return head;
    }
};

File: LeetCode/Explore/linked-list-go/addTwoNumbers.go
package linklist

func addTwoNumbers(l1, l2 *ListNode) *ListNode {
	var head = &ListNode{}
	var p = head
	var carry = 0
	for l1 != nil || l2 != nil || carry != 0 {
		if l1 != nil {
			carry += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			carry += l2.Val
			l2 = l2.Next
		}
		p.Next = &ListNode{Val: carry % 10}
		carry /= 10
		p = p.Next
	}
	return head.Next
}
